presto-main/src/main/java/com/facebook/presto/cost/CachingCostProvider.java

```diff
@@ -53,48 +95,409 @@ public CachingCostProvider(CostCalculator costCalculator, StatsProvider statsPro
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("localMaxMemory", localMaxMemory)
+                    .add("sourcesCosts", sourcesCosts)
+                    .toString();
+        }
+    }
+
+    private static class MemoryEstimate
+    {
+        private final double maxMemoryBeforeOutputting;
+        private final double maxMemoryWhenOutputting;
+
+        public MemoryEstimate(double maxMemoryBeforeOutputting, double maxMemoryWhenOutputting)
+        {
+            this.maxMemoryBeforeOutputting = maxMemoryBeforeOutputting;
+            this.maxMemoryWhenOutputting = maxMemoryWhenOutputting;
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("maxMemoryBeforeOutputting", maxMemoryBeforeOutputting)
+                    .add("maxMemoryWhenOutputting", maxMemoryWhenOutputting)
+                    .toString();
+        }
+    }
+
+    private static class PeakMemoryEstimator
+            extends PlanVisitor<MemoryEstimate, EstimationContext>
+    {
+        @Override
+        protected MemoryEstimate visitPlan(PlanNode node, EstimationContext context)
+        {
+            throw new UnsupportedOperationException("Unsupported plan node type: " + node.getClass().getName());
+        }
+
+        @Override
+        public MemoryEstimate visitRemoteSource(RemoteSourceNode node, EstimationContext context)
+        {
+            return memoryEstimateForStreaming(context);
+        }
+
+        @Override
+        public MemoryEstimate visitAggregation(AggregationNode node, EstimationContext context)
+        {
+            double sourceMaxMemoryWhenOutputting = getOnlySourceMaxMemoryWhenOutputting(node, context);
+            if (node.getStep() == AggregationNode.Step.SINGLE || node.getStep() == AggregationNode.Step.FINAL) {
+                return new MemoryEstimate(
+                        context.localMaxMemory + sourceMaxMemoryWhenOutputting,
+                        context.localMaxMemory);
+            }
+            else {
+                return new MemoryEstimate(
+                        context.localMaxMemory + sourceMaxMemoryWhenOutputting,
+                        context.localMaxMemory + sourceMaxMemoryWhenOutputting);
+            }
+        }
+
+        @Override
+        public MemoryEstimate visitFilter(FilterNode node, EstimationContext context)
+        {
+            return memoryEstimateForStreaming(context);
+        }
+
+        @Override
+        public MemoryEstimate visitProject(ProjectNode node, EstimationContext context)
+        {
+            return memoryEstimateForStreaming(context);
+        }
+
+        @Override
+        public MemoryEstimate visitTopN(TopNNode node, EstimationContext context)
+        {
+            // TODO
+            return new MemoryEstimate(NaN, NaN);
```

Comment:
Does that mean that if there's any `TopNNode` in the plan - the memory estimate will be unknown?

Reply:
If so - i'm not sure how useful the current implementation is. Many nodes cannot be processed based on the memory estimates. So, perhaps we need something similar as explained here: https://github.com/prestodb/presto/pull/11591#issuecomment-431343172. In other words it looks like this code should be inside of the `CostCalculator`. I would only change the trait model a bit

```
 Stats[parent] = FStats(Stats[children]) - nothing changes, that how it is implemented right now
```

```
 Cost[parent] = FCost(Stats[parent], Stats[children], Cost[children])
```

And that should be `CostCalculators` responsibility to compute everything, peak-memory/cpu/network cost. So it looks like the `CostCalculator#calculateCost` interface should accept the `CostProvider` (similarly to StatsProvider). And than based on child costs the peak memory can be estimated.